vertical-pod-autoscaler/pkg/admission-controller/logic/server.go

```diff
@@ -285,22 +70,21 @@ func (s *AdmissionServer) admit(data []byte) (*v1beta1.AdmissionResponse, metric
 		Resource: ar.Request.Resource.Resource,
 	}
 
-	if ar.Request.Resource == podResource {
-		patches, err = s.getPatchesForPodResourceRequest(ar.Request.Object.Raw, ar.Request.Namespace)
-		resource = metrics_admission.Pod
-	} else if admittedGroupResource == vpaGroupResource {
-		patches, err = s.getPatchesForVPADefaults(ar.Request.Object.Raw, ar.Request.Operation == v1beta1.Create)
-		resource = metrics_admission.Vpa
-		// we don't let in problematic VPA objects - late validation
-		if err != nil {
-			status := metav1.Status{}
-			status.Status = "Failure"
-			status.Message = err.Error()
-			response.Result = &status
-			response.Allowed = false
-		}
+	handler, ok := s.resourceHandlers[admittedGroupResource]
+	if ok {
+		patches, err = handler.GetPatches(ar.Request)
+		resource = handler.AdmissionResource()
 	} else {
-		patches, err = nil, fmt.Errorf("expected the resource to be one of: %v, %v", podResource, vpaGroupResource)
+		patches, err = nil, fmt.Errorf("not supported resource type: %v", admittedGroupResource)
+	}
+
+	if handler.DismissIncorrectObjects() && err != nil {
```

Comment:
1. I think this doesn't work when ok is false
2. When we have the handler I think it would be easier to understand if `GetPatches` returned three values - `patches`, `allow` (`bool`), and `err` then here we would have

```
if !allow {
```

Reply:
1. Good catch! Done.
2. Whether to dismiss incorrect objects is a global setting of a handler, not depending on an actual admission request. Eg. Pods should never be dismissed, as we do not own this resource type. So I find it mach cleaner to have it separately. WDYT?